src/DoctrineMessageRepository/DoctrineMessageRepository.php

```diff
@@ -0,0 +1,136 @@
+
+        try {
+            return $this->yieldMessagesFromPayloads($builder->executeQuery()->iterateColumn());
+        } catch (Throwable $exception) {
+            throw UnableToRetrieveMessages::dueTo('', $exception);
+        }
+    }
+
+    /**
+     * @psalm-return Generator<Message>
+     */
+    public function retrieveAllAfterVersion(AggregateRootId $id, int $aggregateRootVersion): Generator
+    {
+        $builder = $this->createQueryBuilder();
+        $builder->where(sprintf('%s = :aggregate_root_id', $this->tableSchema->aggregateRootIdColumn()));
+        $builder->andWhere(sprintf('%s > :version', $this->tableSchema->versionColumn()));
+        $builder->setParameter('aggregate_root_id', $this->uuidSerializer->serializeString($id->toString()));
+        $builder->setParameter('version', $aggregateRootVersion);
+
+        try {
+            return $this->yieldMessagesFromPayloads($builder->executeQuery()->iterateColumn());
+        } catch (Throwable $exception) {
+            throw UnableToRetrieveMessages::dueTo('', $exception);
+        }
+    }
+
+    private function createQueryBuilder(): QueryBuilder
+    {
+        $builder = $this->connection->createQueryBuilder();
+        $builder->select($this->tableSchema->payloadColumn());
+        $builder->from($this->tableSchema->tableName());
+        $builder->orderBy($this->tableSchema->versionColumn(), 'ASC');
+
+        return $builder;
+    }
+
+    /**
+     * @psalm-param iterable<Message> $messages
+     */
+    private function persistMessages(iterable $messages): void
+    {
+        foreach ($messages as $message) {
+            $this->persistMessage($message);
+        }
+    }
+
+    private function persistMessage(Message $message): void
+    {
+        $payload = $this->serializer->serializeMessage($message);
+
+        $payload['headers'][Header::EVENT_ID] ??= Uuid::uuid4()->toString();
+        $payload['headers'][Header::EVENT_ID] = $this->uuidSerializer->serializeString($payload['headers'][Header::EVENT_ID]);
+        $payload['headers'][Header::AGGREGATE_ROOT_ID] = $this->uuidSerializer->serializeString($payload['headers'][Header::AGGREGATE_ROOT_ID]);
+
+        $data = [];
+
+        // Additional columns are added BEFORE known columns, so that additional
+        // columns cannot overwrite (eg) aggregate root ID column.
+        foreach ($this->tableSchema->additionalColumns() as $column => $header) {
+            $data[$column] = $payload['headers'][$header];
+        }
+
+        $data[$this->tableSchema->eventIdColumn()] = $payload['headers'][Header::EVENT_ID];
+        $data[$this->tableSchema->aggregateRootIdColumn()] = $payload['headers'][Header::AGGREGATE_ROOT_ID];
+        $data[$this->tableSchema->versionColumn()] = $payload['headers'][Header::AGGREGATE_ROOT_VERSION];
+        $data[$this->tableSchema->payloadColumn()] = json_encode($payload, $this->jsonEncodeOptions);
+
+        $this->connection->insert($this->tableSchema->tableName(), $data);
```

Comment:
Using `Connection::insert()` eliminates the need to use manually create a query with string interpolation.